lib/services/services.c

```diff
@@ -1079,3 +1080,82 @@ resources_list_agents(const char *standard, const char *provider)
 
     return NULL;
 }
+
+gboolean
+resources_valid_tuple(const char *standard, const char *provider, const char *agent)
+{
+    GList *standards = NULL;
+    GList *providers = NULL;
+    GList *agents = NULL;
+    GListPtr iter = NULL;
+    gboolean rc = FALSE;
+
+    if (crm_str_eq(standard, "stonith", TRUE)) {
+        stonith_t *st = NULL;
+        stonith_key_value_t *devices = NULL;
+        stonith_key_value_t *dIter = NULL;
+        int timeout = 120;
+
+        if (provider != NULL || agent == NULL) {
+            return rc;
+        }
+
+        st = stonith_api_new();
+        st->cmds->list_agents(st, st_opt_sync_call, NULL, &devices, timeout);
+
+        for (dIter = devices; dIter != NULL; dIter = dIter->next) {
+            if (crm_str_eq(dIter->value, agent, TRUE)) {
+                rc = TRUE;
+                break;
+            }
+        }
+
+        stonith_key_value_freeall(devices, 1, 1);
+        stonith_api_delete(st);
+        return rc;
+    }
+
+    standards = resources_list_standards();
+    for (iter = standards; iter != NULL; iter = iter->next) {
+        if (crm_str_eq(iter->data, standard, TRUE)) {
+            rc = TRUE;
+            break;
+        }
+    }
+
+    if (rc == FALSE) {
+        goto done;
+    }
+
+    rc = FALSE;
+
+    if (is_set(pcmk_get_ra_caps(standard), pcmk_ra_cap_provider)) {
+        providers = resources_list_providers(standard);
+        for (iter = providers; iter != NULL; iter = iter->next) {
+            if (crm_str_eq(iter->data, provider, TRUE)) {
+                rc = TRUE;
+                break;
+            }
+        }
+    }
```

Comment:
Maybe overkill, but we could also require provider to be NULL if the standard doesn't support it

Reply:
Yeah, we could.  Are there any standards that support parameters but do not also allow providers?  Otherwise, they're just going to get caught by the test in command line option processing.

Reply:
Good point, yes, both stonith and nagios support parameters but not providers. Stonith of course has its own validate-all command; nagios would just have to be always passed.

Reply:
I think I have this resolved, then.  For some reason I keep getting pretty turned around by this patch series.